include/rocksdb/configurable.h

```diff
@@ -0,0 +1,333 @@
+  // This method assumes that the two objects are of the same class.
+  // @param config_options Controls how the options are compared.
+  // @param other The other object to compare to.
+  // @param mismatch If the objects do not match, this parameter contains
+  //      the name of the option that triggered the match failure.
+  // @param True if the objects match, false otherwise.
+  virtual bool AreEqual(const ConfigOptions& config_options,
+                        const Configurable* other, std::string* name) const;
+
+  // Returns a pretty-printed, human-readable version of the options.
+  // This method is typically used to dump the options to a log file.
+  // Classes should override this method
+  virtual std::string GetPrintableOptions() const { return ""; }
+
+  // Initializes the options and makes sures the settings are valid/consistent
+  // This method may be called as part of Configure (if invoke_prepare_options
+  // is set), or may be invoked separately. Classes must override this method to
+  // provide any implementation-specific initialization, such as opening log
+  // files or setting up cache parameters. Implementations should be idempotent
+  // (e.g. don't re-open the log file or restart the cache), as there is the
+  // potential this method can be called multiple times.
+  //
+  // By default, this method will also prepare all nested (Inner and
+  // OptionType::kConfigurable) objects.
+  //
+  // @param config_options Controls how the object is prepared.  Also contains
+  //      a Logger and Env that can be used to initialize this object.
+  // @return OK If the object was successfully initialized.
+  // @return InvalidArgument If this object could not be successfull
+  // initialized.
+  virtual Status PrepareOptions(const ConfigOptions& config_options);
+
+  // Checks to see if the settings are valid for this object.
+  // This method checks to see if the input DBOptions and ColumnFamilyOptions
+  // are valid for the settings of this object.  For example, an Env might not
+  // support certain mmap modes or a TableFactory might require certain
+  // settings.
+  //
+  // By default, this method will also validate all nested (Inner and
+  // OptionType::kConfigurable) objects.
+  //
+  // @param db_opts The DBOptions to validate
+  // @param cf_opts The ColumnFamilyOptions to validate
+  // @return OK if the options are valid
+  // @return InvalidArgument If the arguments are not valid for the options
+  //       of the current object.
+  virtual Status ValidateOptions(const DBOptions& db_opts,
+                                 const ColumnFamilyOptions& cf_opts) const;
+
+ protected:
+  // True once the object is prepared.  Once the object is prepared, only
+  // mutable options can be configured.
+  bool prepared_;
+  // If this class is a wrapper (has-a), this method should be
+  // over-written to return the inner configurable (like an EnvWrapper).
+  virtual Configurable* Inner() const { return nullptr; }
```

Comment:
If the wrapper is for example 3 level instead of 2 levels (env->EnvWrapper), it will return the inner most object  pointer or just one level inside?

Reply:
Inner should not recurse.  It should only return one-level deep.  If the caller wants to know the deepest level, they would have to implement the recursion themselves.

This method is useful in other cases as well that will come up later.  For example, if passed an Env object and I need to find the Default or Posix one, I could loop over the Inner() ones until I found the one that matched the pattern I need.

Reply:
Maybe better to point out it is not recursive in the comments.